File: src/commands/list-claims.ts
import * as commander from 'commander';
import Table from 'cli-table';

import { actionRunner } from '../utils.js';
import { getAccessToken } from '../config.js';
import { getClaimsList } from '../forma.js';
import VERSION from '../version.js';

const command = new commander.Command();

interface Arguments {
  accessToken?: string;
  filter?: string;
}

command
  .name('list-claims')
  .version(VERSION)
  .description('List claims in your Forma account and their current status')
  .option('--access-token <access_token>', 'Access token used to authenticate with Forma')
  .option(
    '--filter <filter>',
    'Filter claims by status (currently supports: in_progress)',
  )
  .action(
    actionRunner(async (opts: Arguments) => {
      const accessToken = opts.accessToken ?? getAccessToken();

      if (!accessToken) {
        throw new Error(
          "You aren't logged in to Forma. Please run `npx formanator login` first.",
        );
      }

      if (opts.filter && opts.filter !== 'in_progress') {
        throw new Error(
          `Invalid filter value '${opts.filter}'. Currently supported filters: in_progress`,
        );
      }

      const claims = await getClaimsList(
        accessToken,
        opts.filter === 'in_progress' ? 'in_progress' : undefined,
      );

      // Check if any claims have non-null payout_status
      const hasPayoutStatus = claims.some((claim) => claim.payout_status !== null);

      const tableHeaders = [
        'Reimbursement Vendor',
        'Employee Note',
        'Amount',
        'Category',
        'Subcategory',
        'Status',
        'Reimbursement Status',
        ...(hasPayoutStatus ? ['Payout Status'] : []),
        'Date Processed',
        'Note',
      ];

      const table = new Table({
        head: tableHeaders,
      });

      for (const claim of claims) {
        table.push([
          `${claim.reimbursement_vendor}`,
          `${claim.employee_note}`,
          `${claim.amount}`,
          `${claim.category}`,
          `${claim.subcategory}`,
          `${claim.status}`,
          `${claim.reimbursement_status}`,
          ...(hasPayoutStatus ? [`${claim.payout_status}`] : []),
          `${claim.date_processed}`,
          `${claim.note}`,
        ]);
      }

      console.log(table.toString());
    }),
  );

export default command;

File: src/commands/list_claims.rs
use anyhow::{Result, bail};
use tabled::{builder::Builder, settings::Style};

use crate::cli::ListClaimsArgs;
use crate::config::resolve_access_token;
use crate::forma::{ClaimsFilter, get_claims_list};
use crate::verbose;

pub fn run(args: ListClaimsArgs) -> Result<()> {
    verbose::set(args.verbose);
    let access_token = resolve_access_token(args.access_token.as_deref())?;

    let filter = match args.filter.as_deref() {
        None => None,
        Some("in_progress") => Some(ClaimsFilter::InProgress),
        Some(other) => {
            bail!("Invalid filter value '{other}'. Currently supported filters: in_progress")
        }
    };

    let claims = get_claims_list(&access_token, filter)?;

    let has_payout = claims.iter().any(|c| c.payout_status.is_some());

    let mut builder = Builder::default();
    let mut header = vec![
        "Reimbursement Vendor".to_string(),
        "Employee Note".to_string(),
        "Amount".to_string(),
        "Category".to_string(),
        "Subcategory".to_string(),
        "Status".to_string(),
        "Reimbursement Status".to_string(),
    ];
    if has_payout {
        header.push("Payout Status".to_string());
    }
    header.push("Date Processed".to_string());
    header.push("Note".to_string());
    builder.push_record(header);

    for claim in &claims {
        let mut row = vec![
            claim.reimbursement_vendor.clone().unwrap_or_default(),
            claim.employee_note.clone().unwrap_or_default(),
            claim.amount.map(|a| a.to_string()).unwrap_or_default(),
            claim.category.clone().unwrap_or_default(),
            claim.subcategory.clone().unwrap_or_default(),
            claim.status.clone(),
            claim.reimbursement_status.clone().unwrap_or_default(),
        ];
        if has_payout {
            row.push(claim.payout_status.clone().unwrap_or_default());
        }
        row.push(claim.date_processed.clone().unwrap_or_default());
        row.push(claim.note.clone().unwrap_or_default());
        builder.push_record(row);
    }

    let mut table = builder.build();
    table.with(Style::sharp());
    println!("{table}");
    Ok(())
}
